# 4.9 Part I

In [1]:
# Import libarires
import pandas as pd
import numpy as np
import os
import csv

In [2]:
np.set_printoptions(suppress=True)

In [3]:
# Create a path
path = r'C:\Users\SuperHome\Documents\Instacart Basket Analysis'

In [4]:
# Import customers dataframe
cus = pd.read_csv(os.path.join(path,'02 Data','Original Data','customers.csv'), index_col=False)

In [5]:
# Checking dataframe shape
cus.shape

(206209, 10)

## Question 4.
Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.

In [6]:
cus.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [7]:
# Renaming column' First Name' as 'first_name' and 'Surnam' as 'sur_name', 'STATE' as 'State'
cus.rename(columns={'First Name':'first_name','Surnam':'sur_name','STATE':'State'}, inplace = True)

In [8]:
cus

user_id first_name  sur_name  Gender           State  Age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
0                  3    married  165665  
1                  0     single   59285  
2                  2    married   99568  
3                  0     single   42049  
4                  1    married   40374  
...              ...        ...     ...  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095  

[206209 rows x 10 columns]

In [9]:
# Drop column 'data_joined'
df_cus = cus.drop(columns=['date_joined'])

In [10]:
df_cus.shape

(206209, 9)

# Question 5. 
Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data.

In [11]:
df_cus.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [12]:
# Checking nan values
df_cus.isnull().sum()

user_id             0
first_name      11259
sur_name            0
Gender              0
State               0
Age                 0
n_dependants        0
fam_status          0
income              0
dtype: int64

###### As showed above, we have 11259 rows with nan values from 'first_name' column. As I couldn't do anything to fillout them, so I decide to drop these two columns 'frist_name' and 'sur_name' as they won't be used in my analysis project.

In [13]:
# Drop columns 'first_name' and 'sur_name'
cus_new = df_cus.drop(columns=['first_name','sur_name'])

In [14]:
cus_new.shape

(206209, 7)

In [15]:
# Checking mixed-data type
for col in cus_new.columns.tolist():
    weird = (cus_new[[col]].applymap(type) != cus_new[[col]].iloc[0].apply(type)).any(axis = 1)
    if len(cus[weird]) > 0:
        print(col)

###### As we checked there is no mixed-type data find.

In [16]:
# Checking duplicates
cus_new[cus_new.duplicated()]

Empty DataFrame
Columns: [user_id, Gender, State, Age, n_dependants, fam_status, income]
Index: []

###### As checked, there is no duplicate find in this dataframe

In [19]:
cus_new.head(2)

user_id  Gender       State  Age  n_dependants fam_status  income
0    26711  Female    Missouri   48             3    married  165665
1    33890  Female  New Mexico   36             0     single   59285

# Question 6.
Combine your customer data with the rest of your prepared Instacart data. Tip: Make sure the key columns are of the same data type! Hint: Make sure the key columns are the same data type!

In [17]:
# Import dataframe from previous excise
ord_pro = pd.read_pickle(os.path.join(path,'02 Data','Prepared Data','ord_pro_final.pkl'))

In [18]:
ord_pro.head(2)

order_id user_id  order_number  day_of_week  order_hour_of_day  \
0   2539329       1             1            2                  8   
1   2398795       1             2            3                  7   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     7.0         196                  1          0   
1                    15.0         196                  1          1   

  product_name  ...  prices        price_range       Busiest_day  \
0         Soda  ...     9.0  Mid-range product    regularly busy   
1         Soda  ...     9.0  Mid-range product  two_slowest days   

  busiest_period_of_day max_order  loyalty_flag  average_price spending_flag  \
0        Average orders        10  new customer       6.367797   low spender   
1        Average orders        10  new customer       6.367797   low spender   

   median_days_since_prior_order    frequency_flag  
0                           20.0  Regular customer  
1                           20.0  Regular customer  

[2 rows x 22 columns]

In [22]:
# Changing ord_pro dataframe 'user_id'type before merging.
ord_pro['user_id'] = ord_pro['user_id'].apply(int)

In [23]:
cus_merge_ord = cus_new.merge(ord_pro, on = ['user_id'])

In [25]:
# Checking new merged dataframe head
cus_merge_ord.head(2)

user_id  Gender     State  Age  n_dependants fam_status  income  order_id  \
0    26711  Female  Missouri   48             3    married  165665   2543867   
1    26711  Female  Missouri   48             3    married  165665   1285508   

   order_number  day_of_week  ...  prices        price_range  \
0             5            1  ...     9.0  Mid-range product   
1             7            5  ...     9.0  Mid-range product   

        Busiest_day  busiest_period_of_day  max_order  loyalty_flag  \
0  two_busiest days            Most orders          8  new customer   
1    regularly busy            Most orders          8  new customer   

   average_price  spending_flag  median_days_since_prior_order  \
0       7.988889    low spender                           19.0   
1       7.988889    low spender                           19.0   

     frequency_flag  
0  Regular customer  
1  Regular customer  

[2 rows x 28 columns]

# Export new merged dataframe with pickle format

In [26]:
cus_merge_ord.to_pickle(os.path.join(path,'02 Data','Prepared Data','cus_ord_merged_final.pkl'))